### PR TITLE
dialects: Move stencil `CastOp` from experimental to main

### PR DESCRIPTION
**Pull request by AntonLydike**
This PR moves the `CastOp` out of experimental. The CastOp design is unlikely to change, and has full test coverage.

It might make sense to move the `IndexAttr` first, let me know if that is wanted!

### CODE REVIEW

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py [hunk 1]
```diff
@@ -30,6 +29,7 @@
     ExternalLoadOp,
     ExternalStoreOp,
 )
+from xdsl.dialects.stencil import CastOp
```
Comment (georgebisbas) [inline xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:32]: Could move up near lines 15-18

File: xdsl/xdsl_opt_main.py [hunk 2]
```diff
@@ -25,7 +25,8 @@
 from xdsl.dialects.pdl import PDL
 from xdsl.dialects.test import Test
 
-from xdsl.dialects.experimental.stencil import Stencil
+from xdsl.dialects.experimental.stencil import StencilExp
+from xdsl.dialects.stencil import Stencil
```
Comment (georgebisbas) [inline xdsl/xdsl_opt_main.py:29]: same upwards..

**Review by georgebisbas — APPROVED** [review]
Good

File: tests/dialects/test_stencil.py [hunk 3]
```diff
@@ -40,3 +49,103 @@ def test_stencil_return_multiple_ResultType():
...
+
+    # check that math is correct
+    with pytest.raises(VerifyException):
+        cast = CastOp.get(
+            field,
+            IndexAttr.get(-2, -2, -2),
+            IndexAttr.get(100, 100, 100),
+            FieldType.from_shape((100, 100, 100), f32),
+        )
+        cast.verify()
```
Comment (webmiche) [inline tests/dialects/test_stencil.py:74]: Don't you want to check which exceptions are thrown exactly? Or what are these tests supposed to do?

**Review by webmiche — APPROVED** [review]

File: tests/dialects/test_stencil.py (hunk 3 above)
Comment (AntonLydike) [reply tests/dialects/test_stencil.py:74]: How can I check what is thrown exactly? I think it'd be nice to check that!

File: xdsl/xdsl_opt_main.py (hunk 2 above)
Comment (AntonLydike) [reply xdsl/xdsl_opt_main.py]: Done!

File: xdsl/dialects/stencil.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                self.ub,
+                self.result.typ.shape,
+            )
+        ):
+            in_: int = in_attr.value.data
+            out: int = out_attr.value.data
+
+            if ub - lb != out:
+                raise VerifyException(
+                    "Bound math doesn't check out in dimensions {}! {} - {} != {}".format(
```
Comment (superlopuh) [inline xdsl/dialects/stencil.py:85]: nit: f strings

**Review by math-fehr — APPROVED** [review]
:+1:

**Pull request merged**